### ISSUE TITLE
allow running non-captured code during execution for flexible client/server benchmarking

### ISSUE THREAD

**Issue by hargut** [iai-callgrind/iai-callgrind#242]
Especially for server/client application benchmarking it would be helpful to be able to execute the client or server code-path in a non-monitored way.

As of now could not find a way to handle that scenario with separation between client/server for benchmarking the server application only. I was able to manage benchmarking the server & client at the same time. For the intended use-case it would be beneficial to have a clear analysis/comparision of either server or client only.

Currently the framework supports:
- setup()
   - which is run before the application starts
- teardown()
  - which is run after the application has run

Suggest to extend the options to:
- setup()
- action()
  - which is run after the application process is spawned
  - could issue signals to stop the benchmarked process after successful action execution
- teardown()

This would allow for example a setup like:
```rust
fn run_benchmark_requests() {
    tokio::runtime::Builder::new_multi_thread()
        .enable_all()
        .build()
        .unwrap()
        .block_on(async {
            println!("Executing benchmark requests...");
            wait_for_tcp_connect(Version::HTTP_11).await;
            println!("TCP connect successful.");
            tls_post_data(Version::HTTP_11, Some(&CLIENT_HTTP11), generate_request_bodies(64)).await;
            println!("Sent benchmark requests successfully.");
        })
}

#[binary_benchmark]
#[bench::http_11_requests(action = run_benchmark_requests)]
fn acceptor_testing() -> Command {
    let path = format!("{}/../target/release/examples/echo_server", env!("CARGO_MANIFEST_DIR"));
    Command::new(path).build()
}
```

Similar this could be used for benchmarking client that connects to a server.
- spawning the client process within valgrind
  - having it wait for the server connection
- running the server as action
- client is benchmarked

This use of a generic `action` would as well allow for other tests where it is desirable to execute any uncovered code while running the benchmark.

I've already started working on a patch that supports the scenario covering a server application, and would be willing to continue working on it in case the approach is supported.

For any other suggestions/ideas/layouts to support a server/client scenario I'm as well open.
Please let me know your thoughts on the topic or eventually existing alternative approaches.

Many thanks in advance & have a nice weekend. :grinning: 

Kind regards,
Harald

**Comment by gamma0987** [comment]
Hi Harald.

Thanks for your detailed description of the problem.

> As of now could not find a way to handle that scenario with separation between client/server for benchmarking the server application only. I was able to manage benchmarking the server & client at the same time.

If you managed to benchmark the server and client at the same time, you should have a closer look at the callgrind option [`--toggle-collect`](https://valgrind.org/docs/manual/cl-manual.html#cl-manual.options.collection). The `BinaryBenchmarkConfig` allows you to pass arguments to callgrind for example with

```rust
BinaryBenchmarkConfig::default().raw_callgrind_args(["--toggle-collect=echo_server::main"])
```

and if you want to benchmark the client

```rust
BinaryBenchmarkConfig::default().raw_callgrind_args(["--toggle-collect=client::main"])
```

You can also use valgrind client requests (more details are in the section about client requests in the [README](https://github.com/iai-callgrind/iai-callgrind?tab=readme-ov-file#valgrind-client-requests)). For example in your server:

```rust
fn main() {
    iai_callgrind::client_requests::callgrind::start_instrumentation();
    /* start the server ... */
    iai_callgrind::client_requests::callgrind::stop_instrumentation();
}
```

and tell callgrind to not start event counting at the start of the executable with `--instr-atstart=no`.

Please give me your feedback to the suggestions above. I hope this helps in setting up your use case.

I can't tell right now, if having an `action` is the right solution to the problem. Maybe the same can achieved with the `setup` and `teardown` hooks which are already in place. I'm currently a little bit preoccupied with #239 but I will have a closer look at it soon. If you want to work on a solution, you're most welcome. Just please until there is some common ground to not waste your efforts.

Just some quick thoughts. Instead of an `action`, what about having the possibility to start the `setup` and `Command` in parallel as it is already done when the `Command` is expecting the output of the `setup` as its `Stdin`? This behaviour is also described in more detail in the README.

I had some rough thoughts about a `Command::delay` function when I wrote on the new binary benchmark api. This funcion can be used to synchronise the execution of the `setup` and `Command` if they are executed in parallel. For example delay the execution of the `Command` until a specific event in the `setup` happened like the creation of a file. That should at least help to benchmark a client, since you can wait for the server, which is started in the `setup`, to drop a pid file or something similar when it is ready. As soon as the pid file is there the client `Command` is executed and benchmarked. You can stop the server in `teardown` for example by sending SIGHUP to the pid from the pidfile.

**Comment by hargut** [comment]
Hi @gamma0987,

thank you very much for fast response, your input and thoughts on that topic. I highly appreciate that. :+1: 
I think that some of the provided alternative approaches could solve, or even enhance the intended test scenario. 

Especially the `client_requests` could be very helpful. I've seen it in the docs, but somehow skipped the details on it.

Thank you for all your efforts on `iai_callgrind`, it is a really helpful tool.

Have a great weekend. :sunflower: 

Kind regards,
Harald

**Comment by gamma0987** [comment]
Hi Harald,

glad to help out :) Thanks for your positive feedback. By the way the [new guide](https://iai-callgrind.github.io/iai-callgrind) is online and I it hope it is clearer than the old README.

Best, gamma

**Comment by gamma0987** [comment]
I have thought about an `action` but I don't think it is necessary. Having the possibility to execute the `setup` in parallel with the `Command`, should solve the issues you had. 

If you're interested in a solution for testing the server with the current Iai-Callgrind implementation, here's what I quickly set up misusing the `Command::stdin(Stdin::Setup(Pipe::Stdout))` to force the parallel execution of `setup` and `Command`. This is not ideal, but as long as the `server` doesn't read from `Stdin` and the `setup` function doesn't output anything to `Stdout` it's ok to do it as below.

```rust
use std::fs::File;
use std::thread::sleep;
use std::time::Duration;

use iai_callgrind::{
    binary_benchmark, binary_benchmark_group, main, BinaryBenchmarkConfig, Pipe, Sandbox, Stdin,
    Stdio,
};

fn setup() {
    let pid_file = std::path::PathBuf::from("server.pid");
    while !pid_file.exists() {
        sleep(Duration::from_millis(200));
    }

    std::fs::write("some_text", b"some_text").unwrap();

    _ = std::process::Command::new("nc")
        .arg("127.0.0.1")
        .arg("12345")
        .arg("-N")
        .stdin(File::open("some_text").unwrap())
        .status()
        .unwrap();

    let pid = std::fs::read("server.pid").unwrap();
    _ = std::process::Command::new("kill")
        .arg("-15")
        .arg(std::str::from_utf8(&pid).unwrap())
        .status()
        .unwrap();
}

#[binary_benchmark(
    config = BinaryBenchmarkConfig::default().sandbox(Sandbox::new(true)),
    setup = setup()
)]
fn bench_binary() -> iai_callgrind::Command {
    iai_callgrind::Command::new(env!("CARGO_BIN_EXE_server"))
        .arg("--bind=127.0.0.1")
        .arg("--port=12345")
        .arg("--pid-file=server.pid")
        .stdin(Stdin::Setup(Pipe::Stdout))
        .stdout(Stdio::Inherit)
        .stderr(Stdio::Inherit)
        .build()
}

binary_benchmark_group!(name = my_group; benchmarks = bench_binary);
main!(binary_benchmark_groups = my_group);
```

The server I've used was a simple echo server dropping a `pidfile` as soon as it is ready and listening for the signal SIGTERM.

I would also love to be able to delay the execution of the `Command` as I've described in my previous comment. So, testing a client should also be possible.

Are you still interested in implementing this?

**Comment by hargut** [comment]
Hi @gamma0987,

thank you for your input & example. I've done some tests today and think that a solution that keeps `setup` running during the binary execution will work for this use case.

For the client side this works as well:
- start the server from the `setup` with a pid file
- run the client binary under benchmark
- once tests are done, stop the server in a `teardown`

I've also tested the work-around which keeps the setup running as required and could successfully run some benchmarks.

During tests with `iai_callgrind::client_requests::callgrind::start_instrumentation() & stop_instrumentation()` I noticed that the comparisons do not work, and crated a fix for that within #252. Valgrind Version 3.22.0 was in use.

Initially I intended to use the `client_requests` with a LibraryBenchmark. So I did some tests on that, but for some reason I could not manage to get usable callgraphs from valgrind through spawned threads.
The same looks to be working fine when using a BinaryBenchmark. The resulting callgraphs have the thread spawning included, and after that several libc/low-level calls are visible, but none of them can be easily correlated to the code. I assume that this is unrelated to iai_callgrind and likely a valgrind user-error. Did you already see such situation or have an idea what the cause could be?

Thank you for your support. :smiley: 

Kind regards,
Harald

**Comment by hargut** [comment]
As I'm interested in helping to implement the feature I've looked into the workaround relevant parts within Assistant and the Benchmark:
https://github.com/iai-callgrind/iai-callgrind/blob/96bcfdc86e97aa07f54dd3e5eada7cc47dbb7c23/iai-callgrind-runner/src/runner/bin_bench.rs#L191-L204
https://github.com/iai-callgrind/iai-callgrind/blob/96bcfdc86e97aa07f54dd3e5eada7cc47dbb7c23/iai-callgrind-runner/src/runner/common.rs#L129-L139

How would you like to have the option for `setup()` parallel run be implemented?
Should it be part of the `Command` configuration, e.g. `wait_for_setup(false)`?
Which kind of signals (file creation, network connect, ...) would be desired for the `Command::delay`?

**Comment by hargut** [comment]
Probably a `delay` function could be a versatile solution. Once the function is successfully finished, the benchmark runs.

**Comment by gamma0987** [comment]
Thanks for the fix, I'll have a look at it soon. Concerning the tread problem, can you post a minimal example with the crates (tokio?) you have used? I need more context to be able to have a look into it. Have you tried valgrind options like `--fair-shed`?

> Should it be part of the Command configuration, e.g. wait_for_setup(false)?

Yes, I think `Command` is a good start, maybe adding it to the `BinaryBenchmarkConfig`, too. `Command` in the `iai_callgrind_runner::api` stores a `BinaryBenchmarkConfig` so adding it to both is really easy. I'm not sure about the name though. But `wait_for_setup` sounds wrong since we don't wait for the setup. More something into the direction of `Command::setup_is_parallel`.

> How would you like to have the option for setup() parallel run be implemented?

Integrate it as far as possible with what is already there in the code part you have already outlined. That shouldn't be much I guess. The `Stdin::Setup` setting of `Command::stdin` takes precedence over the setting in `setup_is_parallel`.

> Which kind of signals (file creation, network connect, ...) would be desired for the Command::delay?

Checking connectivity to a socket is a great idea. File existence definitely and also the most basic one to wait for a `Duration` worth of time. `Command::delay` should take a struct or enum `Delay`, so we can easily extend the functionality. Also, it would be great to implement `From<Duration> for Delay` etc. for an easier usage.

I think there should be a timeout (just a `Duration`) setting for `Delay`, too. To not  destroy the simple usage per `From<???> for Delay`, having a default of around `10 Minutes` would be great. Just enough to avoid hanging the CI forever if something goes wrong. The polling time should also be adjustable maybe with a default of `10ms`.

**Comment by hargut** [comment]
After having some more tests in the area of using `--toggle-collect` with BinaryBenchmarks this looks to be an interesting option, as it allows to focus the results on server functions without having to touch the codebase.

What I noticed during several tests is that the mechanism to run LibraryBenchmarks has multiple restrictions compared to BinaryBechmarks for the use case. The `--toggle-collect=*::__iai_callgrind_wrapper_mod::*` option for the valgrind command is the cause for the missing thread details.
Adding another `--toggle-collect=function::identifier` allows to collect the data, and have it visible in the call-graph. But in that case the benchmark results are not comparable with the ones produced by BinaryBenchmarks as the SentinelParser stops reading the call-graph after having detected the sentinel, and therefore misses all call statistics from the threads.

https://github.com/iai-callgrind/iai-callgrind/blob/492e1a1df4e1aaf0f256463354369f8f1ed4ab86/iai-callgrind-runner/src/runner/mod.rs#L36
https://github.com/iai-callgrind/iai-callgrind/blob/492e1a1df4e1aaf0f256463354369f8f1ed4ab86/iai-callgrind-runner/src/runner/lib_bench.rs#L329
https://github.com/iai-callgrind/iai-callgrind/blob/492e1a1df4e1aaf0f256463354369f8f1ed4ab86/iai-callgrind-runner/src/runner/lib_bench.rs#L153
https://github.com/iai-callgrind/iai-callgrind/blob/492e1a1df4e1aaf0f256463354369f8f1ed4ab86/iai-callgrind-runner/src/runner/callgrind/sentinel_parser.rs#L54-L65


To reproduce the issue the following can be used:
```rust
use std::thread;
use std::time::Duration;
use iai_callgrind::{main, library_benchmark, library_benchmark_group, LibraryBenchmarkConfig };

fn child_call() {
    let handle = thread::spawn(|| {
        println!("From the child.");
        thread::sleep(Duration::from_millis(500));
    });
    handle.join().expect("TODO: panic message");
}

#[library_benchmark]
#[bench::visualize_graph()]
fn calling_child() {
    println!("From the parent.");
    child_call();
}

library_benchmark_group!(
    name = sentinel_parser;
    config = LibraryBenchmarkConfig::default()
        .raw_callgrind_args([
            ""
            , "--toggle-collect=*"
    ]);
    benchmarks = calling_child,
);

main!(
    library_benchmark_groups = sentinel_parser
);
```
The results even drop by approx 40% when just adding the `--toggle-collect=*` to the valgrind args due to changed order within the callgraph file.

After having tested, learned and read a bunch on the topic of valgrind benchmarking rust code with iai_callgrind for now my approach to the current problem will be to use:
- BinaryBenchmarks
- with additional `--toggle-collect`

**Comment by gamma0987** [comment]
Thanks for the details. This helps a lot. But, I think I have to clearify things a little bit.

The sentinel parser is doing the right thing. It parses the benchmark function(s) with the default toggle and collects all the metrics in it.

To get to the root of the problem, I ran your benchmark from above but without the `config` in the `library_benchmark_group` setting an additional toggle. The metrics include all the thread spawns in the benchmark function and what is executed in them. 

```text
test_lib_bench_dev::sentinel_parser::calling_child visualize_graph
  Instructions:                4557|N/A             (*********)
  L1 Hits:                     6134|N/A             (*********)
  L2 Hits:                      110|N/A             (*********)
  RAM Hits:                     237|N/A             (*********)
  Total read+write:            6481|N/A             (*********)
  Estimated Cycles:           14979|N/A             (*********)
```

```
.          .          .          .          .          .          .          .          .           use std::thread;
.          .          .          .          .          .          .          .          .           use std::time::Duration;
.          .          .          .          .          .          .          .          .
.          .          .          .          .          .          .          .          .           use iai_callgrind::{library_benchmark, library_benchmark_group, main, LibraryBenchmarkConfig};
.          .          .          .          .          .          .          .          .
.          .          .          .          .          .          .          .          .           fn child_call() {
1 ( 0.02%) 0          1 ( 0.12%) .          .          .          .          .          .               let handle = thread::spawn(|| {
3,545 (77.79%) 844 (75.22%) 575 (71.70%) 193 (71.22%) 32 (76.19%) 34 (100.0%) 122 (61.00%) 9 (64.29%) 23 (100.0%)  => /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/thread/mod.rs:std::thread::spawn (1x)
.          .          .          .          .          .          .          .          .                   println!("From the child.");
.          .          .          .          .          .          .          .          .                   thread::sleep(Duration::from_millis(500));
.          .          .          .          .          .          .          .          .               });
6 ( 0.13%) 2 ( 0.18%) 3 ( 0.37%) .          .          .          .          .          .               handle.join().expect("TODO: panic message");
538 (11.81%) 152 (13.55%) 107 (13.34%) 34 (12.55%) 1 ( 2.38%) 0          34 (17.00%) .          .           => /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/thread/mod.rs:std::thread::JoinHandle<T>::join (1x)
.          .          .          .          .          .          .          .          .           }
.          .          .          .          .          .          .          .          .
5 ( 0.11%) 2 ( 0.18%) 1 ( 0.12%) 1 ( 0.37%) 0          0          1 ( 0.50%) .          .           #[library_benchmark]
4,557 (100.0%) 1,122 (100.0%) 802 (100.0%) 271 (100.0%) 42 (100.0%) 34 (100.0%) 200 (100.0%) 14 (100.0%) 23 (100.0%)  => benchmark-tests/benches/test_lib_bench/dev/dev.rs:test_lib_bench_dev::calling_child::__iai_callgrind_wrapper_mod::calling_child (1x)
.          .          .          .          .          .          .          .          .           #[bench::visualize_graph()]
.          .          .          .          .          .          .          .          .           fn calling_child() {
1 ( 0.02%) 1 ( 0.09%) 1 ( 0.12%) 0          1 ( 2.38%) .          .          .          .               println!("From the parent.");
450 ( 9.87%) 121 (10.78%) 109 (13.59%) 42 (15.50%) 8 (19.05%) 0          42 (21.00%) 5 (35.71%) .           => /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/io/stdio.rs:std::io::stdio::_print (1x)
.          .          .          .          .          .          .          .          .               child_call();
.          .          .          .          .          .          .          .          .           }
.          .          .          .          .          .          .          .          .
.          .          .          .          .          .          .          .          .           library_benchmark_group!(
.          .          .          .          .          .          .          .          .               name = sentinel_parser;
.          .          .          .          .          .          .          .          .               // config = LibraryBenchmarkConfig::default() .raw_callgrind_args([ "" ,
.          .          .          .          .          .          .          .          .               //     "--toggle-collect=*" ]);
.          .          .          .          .          .          .          .          .               benchmarks = calling_child,
```

Note setting this additional toggle `--toggle-collect=*` (the default toggle is still active) produces the wrong metrics (the by 40% dropped ones) although now you can see the calls within the spawned thread of the benchmark function in the call-graph. To be able to tell whether the metrics are actually correct, I ran the benchmark again but now I wasn't not adding the default toggle to the callgrind executable arguments. I had to do this in the source code because there is currently no option to unset the default toggle or use a custom one (I'm going to try to better the situation in #254). Running the benchmark again without the default toggle, the metrics were the same compared to running the benchmark with default toggle.

```text
test_lib_bench_dev::sentinel_parser::calling_child visualize_graph
  Instructions:                4557|4557            (No change)
  L1 Hits:                     6134|6134            (No change)
  L2 Hits:                      110|110             (No change)
  RAM Hits:                     237|237             (No change)
  Total read+write:            6481|6481            (No change)
  Estimated Cycles:           14979|14979           (No change)
```

However, what has changed is, that now the visible call-graph included what is happening inside the thread from the benchmark function. 

```text
.          .          .          .          .          .          .          .          .           use std::thread;
.          .          .          .          .          .          .          .          .           use std::time::Duration;
.          .          .          .          .          .          .          .          .
.          .          .          .          .          .          .          .          .           use iai_callgrind::{library_benchmark, library_benchmark_group, main, LibraryBenchmarkConfig};
.          .          .          .          .          .          .          .          .
.          .          .          .          .          .          .          .          .           fn child_call() {
1 ( 0.00%) 0          1 ( 0.00%) .          .          .          .          .          .               let handle = thread::spawn(|| {
3,545 ( 0.91%) 844 ( 0.90%) 575 ( 1.33%) 193 ( 8.54%) 32 ( 1.25%) 34 ( 3.17%) 122 ( 5.90%) 9 ( 0.51%) 23 ( 2.33%)  => /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/thread/mod.rs:std::thread::spawn (1x)
1 ( 0.00%) 1 ( 0.00%) 1 ( 0.00%) .          .          .          .          .          .                   println!("From the child.");
370 ( 0.10%) 99 ( 0.11%) 86 ( 0.20%) 13 ( 0.57%) 3 ( 0.12%) 1 ( 0.09%) 2 ( 0.10%) 1 ( 0.06%) 1 ( 0.10%)  => /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/io/stdio.rs:std::io::stdio::_print (1x)
3 ( 0.00%) 1 ( 0.00%) 1 ( 0.00%) .          .          .          .          .          .                   thread::sleep(Duration::from_millis(500));
116 ( 0.03%) 24 ( 0.03%) 20 ( 0.05%) 7 ( 0.31%) 1 ( 0.04%) 0          7 ( 0.34%) .          .           => /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/thread/mod.rs:std::thread::sleep (1x)
.          .          .          .          .          .          .          .          .               });
6 ( 0.00%) 2 ( 0.00%) 3 ( 0.01%) .          .          .          .          .          .               handle.join().expect("TODO: panic message");
538 ( 0.14%) 152 ( 0.16%) 107 ( 0.25%) 34 ( 1.50%) 1 ( 0.04%) 0          34 ( 1.64%) .          .           => /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/thread/mod.rs:std::thread::JoinHandle<T>::join (1x)
.          .          .          .          .          .          .          .          .           }
.          .          .          .          .          .          .          .          .
9 ( 0.00%) 4 ( 0.00%) 3 ( 0.01%) 2 ( 0.09%) 0          0          1 ( 0.05%) .          .           #[library_benchmark]
4,557 ( 1.18%) 1,122 ( 1.20%) 802 ( 1.85%) 271 (11.99%) 42 ( 1.64%) 34 ( 3.17%) 200 ( 9.67%) 14 ( 0.79%) 23 ( 2.33%)  => benchmark-tests/benches/test_lib_bench/dev/dev.rs:test_lib_bench_dev::calling_child::__iai_callgrind_wrapper_mod::calling_child (1x)
.          .          .          .          .          .          .          .          .           #[bench::visualize_graph()]
.          .          .          .          .          .          .          .          .           fn calling_child() {
1 ( 0.00%) 1 ( 0.00%) 1 ( 0.00%) 0          1 ( 0.04%) .          .          .          .               println!("From the parent.");
450 ( 0.12%) 121 ( 0.13%) 109 ( 0.25%) 42 ( 1.86%) 8 ( 0.31%) 0          42 ( 2.03%) 5 ( 0.28%) .           => /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/io/stdio.rs:std::io::stdio::_print (1x)
.          .          .          .          .          .          .          .          .               child_call();
.          .          .          .          .          .          .          .          .           }
.          .          .          .          .          .          .          .          .
.          .          .          .          .          .          .          .          .           library_benchmark_group!(
.          .          .          .          .          .          .          .          .               name = sentinel_parser;
.          .          .          .          .          .          .          .          .               // config = LibraryBenchmarkConfig::default() .raw_callgrind_args([ "" ,
.          .          .          .          .          .          .          .          .               //     "--toggle-collect=*" ]);
.          .          .          .          .          .          .          .          .               benchmarks = calling_child,
```

In summary, the shown and parsed metrics of Iai-Callgrind seem to be correct but for some reason the call-graph doesn't show what is happening inside the threads as long as the default toggle is active.

> After having tested, learned and read a bunch on the topic of valgrind benchmarking rust code with iai_callgrind for now my approach to the current problem will be to use ...

I'm glad you found a simple solution that works.

**Comment by gamma0987** [comment]
Short addition after having a closer look at it:

The root of the problem seems to be, that callgrind separates the threads internally and only shows all of them with the `--separate-threads=yes` option. Also, the default toggle is applied to each thread not only the main thread and the default-toggle also sets `collect-at-start=no`.

Currently, the `--separate-threads` option is disabled by iai-callgrind because it would create a separate callgrind output file per thread but iai-callgrind can currently handle only a single callgrind output file so `--combine-dumps=yes` is also set. iai-callgrind parses the combined dump files correctly but, `callgrind_annotate` can't handle combined dumps. I think, a solution to the problem (in addition to #254) would be to allow and parse multiple output files instead of combining them per default.

**Comment by gamma0987** [comment]
Closing this as completed with #256, #255
